numpy/core/arrayprint.py

```diff
@@ -1455,7 +1455,14 @@ def array_str(a, max_line_width=None, precision=None, suppress_small=None):
     # so floats are not truncated by `precision`, and strings are not wrapped
     # in quotes. So we return the str of the scalar value.
     if a.shape == ():
-        return str(a[()])
+        item = a[()]
+        # For "normal" ndarrays we want to return just "str(item)", but many
+        # ndarray subclasses wrap any scalar output back into the subclass,
+        # which would cause infinite recursion. In those cases, return
+        # "str(item.item())" instead.
+        if type(a) is np.ndarray or not isinstance(item, type(a)):
+            return str(item)
```

Comment:
I think this justifies a `_get_true_scalar(arr_0d)` helper function, which you implement as
```
def _get_true_scalar(arr_0d)
    item = a[()]
    if type(a) is np.ndarray or not isinstance(item, type(a)):
        return item
    return item.item()
```
I'm not sure this implementation is ideal though - should the last line be `a.view(np.ndarray)[()]` instead? Or `a.item()`?